File: app/src/main/java/com/example/engineeraitest/data/repository/UserRepositoryImpl.java
package com.example.engineeraitest.data.repository;

import com.example.engineeraitest.data.api.UserApiService;
import com.example.engineeraitest.data.api.mapper.UserMapper;
import com.example.engineeraitest.domain.UserRepository;
import com.example.engineeraitest.domain.model.User;
import com.jakewharton.retrofit2.adapter.rxjava2.HttpException;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;

public class UserRepositoryImpl implements UserRepository {

    private UserApiService mUserApiService;

    public UserRepositoryImpl(UserApiService mUserApiService) {
        this.mUserApiService = mUserApiService;
    }

    @Override
    public Single<List<User>> getUsers(int offset, int limit) {
        return mUserApiService.getUsers(offset, limit)
                .map(userResponse -> {
                    if (!userResponse.isSuccessful() || !userResponse.body().getStatus()) {
                        throw new HttpException(userResponse);
                    }
                    return userResponse.body().getData().getUsers();
                })
                .toFlowable()
                .flatMap(Flowable::fromIterable)
                .map(new UserMapper())
                .toList();
    }
}

File: app/src/main/java/com/example/engineeraitest/data/repository/UserRepositoryImpl.kt
package com.example.engineeraitest.data.repository

import com.example.engineeraitest.data.api.UserApiService
import com.example.engineeraitest.data.api.mapper.UserMapper
import com.example.engineeraitest.domain.exception.UserRepository
import com.example.engineeraitest.domain.model.User
import com.jakewharton.retrofit2.adapter.rxjava2.HttpException
import io.reactivex.Flowable
import io.reactivex.Single

class UserRepositoryImpl(private val mUserApiService: UserApiService) : UserRepository {

    override fun getUsers(offset: Int?, limit: Int?): Single<List<User>> {
        return mUserApiService.getUsers(offset, limit)
            .map {
                if (!it.isSuccessful) throw HttpException(it)
                it.body()?.data?.users!!
            }
            .toFlowable()
            .flatMap { Flowable.fromIterable(it) }
            .map(UserMapper())
            .toList()
    }
}